### DISCUSSION TITLE
RmlUi 4.0

### DISCUSSION THREAD

**Discussion by mikke89** [mikke89/RmlUi#187] [General]
RmlUi 4.0 comes densely packed with several highly requested and impactful new features, as well as many bug fixes. 

### Notable new features

- [Data binding](https://mikke89.github.io/RmlUiDoc/pages/data_bindings.html) using a model-view-controller (MVC) approach. A built-in feature for synchronizing the user interface with the application data.
- New RCSS properties: `border-radius`, `word-break`, `box-sizing`, `caret-color`.
- RCSS table support.
- Media queries support.
- Improved high DPI support with `resolution` media feature and sprite sheet target scaling.
    - Samples now implement responsive high DPI support on Windows 10.
- New elements: `<label>`, `<svg>` (with plugin), `<lottie>` (with plugin).
- Added `Element::QuerySelector`, `Element::QuerySelectorAll`, and `Element::Closest`.
- Added `ElementDocument::ReloadStyleSheet` to reload styles without affecting the document tree.
- Several layout improvements including better overflow clipping and shrink-to-fit widths.
- Many improvements to the Lua plugin, as well as [detailed documentation](https://mikke89.github.io/RmlUiDoc/pages/lua_manual.html).

Read the [full changelog here](https://github.com/mikke89/RmlUi/blob/master/changelog.md#rmlui-40), the above is just a small selection of changes.

The new features have also been thoroughly documented, so make sure to read the [official documentation](https://mikke89.github.io/RmlUiDoc/) for details. RmlUi has been restructured to simplify its usage, resulting in larger breaking changes. Users upgrading from RmlUi 3.x are encouraged to read the [upgrade guide](https://github.com/mikke89/RmlUi/blob/master/changelog.md#restructuring-rmlui).

Finally, it has been fun to see the community grow over the last year, and I'm very happy that you are all helpful and welcoming to members both new and old. It has been a delight to see the uptake in contributions from users. Thank you for all the contributions and let's keep it going!

Windows binaries built using Visual Studio 15 2017.

<hr /><em>This discussion was created from the release <a href='https://github.com/mikke89/RmlUi/releases/tag/4.0'>RmlUi 4.0</a>.</em>

**Comment by Dakror** [comment]
Hooray! Thanks for your continued efforts on this amazing library!

**Reply by mikke89** [reply]
Thank you! :)

**Reply by viciious** [reply]
I second that :)

**Reply by aquawicket** [reply]
Awesome. a very pleasant surprise indeed :)

**Comment by Division** [comment]
Thank you for your work! Writing a render backend doesn't seem too complicated so I'm planning to use RmlUi in my hobby engine soon!

**Reply by mikke89** [reply]
Thanks! Hope you have a good experience with the library.
